### PR TITLE
Do not unwrap when creating a bucket

### PR DESCRIPTION
**Pull request by irevoire**
Hey, while trying this amazing lib, I miss-typed my URL and got an unwrap from the lib; I guess it’s better to forward the error to the caller than panicking in the callee, right?

### CODE REVIEW

**Comment by paolobarbolini** [comment]
Thanks for the PR. We didn't expect that `unwrap` to be reachable honestly. Could you share the mistyped URL?

**Comment by irevoire** [comment]
Hey @paolobarbolini,

Here is the exact code I ran to trigger this panic:
```rust
use std::time::Duration;

use rusty_s3::actions::{CreateBucket, S3Action};
use rusty_s3::{Bucket, Credentials, UrlStyle};

#[test]
pub fn create_bucket() {
    let mut buf = [0; 8];
    getrandom::getrandom(&mut buf).expect("getrandom");

    let name = "tamo".to_string();

    let url = "http://localhost:9000".parse().unwrap();
    let key = "minioadmin";
    let secret = "minioadmin";
    let region = "minio";

    let bucket = Bucket::new(url, UrlStyle::VirtualHost, name, region).unwrap();
    let credentials = Credentials::new(key, secret);

    let action = CreateBucket::new(&bucket, &credentials);
    let url = action.sign(Duration::from_secs(60));
    ureq::put(dbg!(url.as_str())).call().unwrap();
}
```

And here’s the logs:
```
[tests/create_bucket.rs:23] url.as_str() = "http://tamo.localhost:9000/?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minioadmin%2F20230926%2Fminio%2Fs3%2Faws4_request&X-Amz-Date=20230926T171833Z&X-Amz-Expires=60&X-Amz-SignedHeaders=host&X-Amz-Signature=29a6c757fb080c06dd0834df2765b3004592aee46e47c700dd1712be11b3c0f6"
thread 'create_bucket' panicked at tests/create_bucket.rs:23:42:
called `Result::unwrap()` on an `Err` value: Transport(Transport { kind: Dns, message: Some("resolve dns name 'tamo.localhost:9000'"), url: Some(Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("tamo.localhost")), port: Some(9000), path: "/", query: Some("X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minioadmin%2F20230926%2Fminio%2Fs3%2Faws4_request&X-Amz-Date=20230926T171833Z&X-Amz-Expires=60&X-Amz-SignedHeaders=host&X-Amz-Signature=29a6c757fb080c06dd0834df2765b3004592aee46e47c700dd1712be11b3c0f6"), fragment: None }), source: Some(Custom { kind: Uncategorized, error: "failed to lookup address information: nodename nor servname provided, or not known" }) })
```

So, from what I understand,
> Could you share the miss-typed URL?

The URL is `http://tamo.localhost:9000/?...stuff...`.

Also, the fix is to change the `UrlStyle::VirtualHost` to `UrlStyle::Path`.

I’ll be honest, I’m a huge S3 noob, so I have no idea of what it means (I see it changes the URL, but I don't even know why I have the choice 😅).

You can reproduce this issue easily on your side by updating the `tests/common.rs` file and changing the `UrlStyle`. Then just run `cargo test --test create_delete_bucket`

**Review by guerinoni — APPROVED** [review]

**Review by paolobarbolini — APPROVED** [review]
Thanks for the fix and the for posting the reproduction. I'll try merging and releasing over the weekend

File: src/bucket.rs [hunk 1]
```diff
@@ -311,6 +318,7 @@ impl Display for BucketError {
         match self {
             Self::UnsupportedScheme => f.write_str("unsupported Url scheme"),
             Self::MissingHost => f.write_str("Url is missing the `host`"),
+            Self::ParseError(e) => f.write_str(&e.to_string()),
```
Comment (paolobarbolini) [inline src/bucket.rs:321]: Something like this should be a bit nicer and avoid an heap allocation

```suggestion
            Self::ParseError(e) => e.fmt(f),
```

**Pull request merged**